### ISSUE TITLE
Complete range search API

### ISSUE THREAD

**Issue by eightysteele** [eightysteele/Spatial-Data-Library#19]
See: https://github.com/eightysteele/Spatial-Data-Library/wiki/REST-API

**Comment by eightysteele** [comment]
We're moving ahead with range indexes on App Engine for the following variables:
- mean temp (BIO1)
- alt
- precipitation (BIO12)

**Comment by eightysteele** [comment]
@tucotuco: Can you put together a CSV for all Worldclim layer variable metadata? We need that for the variables API. Maybe we can ping RJH for this if he has it (to save us a little work)? 

Update: Created new issue #29 for this.

**Comment by eightysteele** [comment]
@tucotuco, @vdave: What range intervals make sense for `mean temp`, `altitude`, and `precipitation`? We need 3 range intervals for each (like within 1, within 10, within 50 degrees of `mean temp` for example).

**Comment by tucotuco** [comment]
Hmm. Interesting issue. Worldclim temperatures are given to 0.1 degree precision, though stored as decidegrees (142 in the data store is 14.2 degrees). A range query for within-1-degree of 14.2 should return records for any values in the datastore in the list [132,133,134,...142...150,151,152]. That's 21 values rather than the 3 we expected when we first thought of the solution as [141,142,143]. That's an App Engine storage cost of $21.90 per month for the 12 tmean variables using the per-character cost of $0.029 per month) for only the 1-degree range.

What about this alternative?

Instead of within-ranges centered on the actual value of the variable, what if we used predefined absolute ranges? Suppose the lower and upper limits of temperature are -19.4 and 40.6. One set of categories would divide the range into 1-degree intervals:

[-20, -19}, [-19,-18}, [-18,-17}...[39,40},[40,41}  (61 intervals)

The interval would be stored as its lower bound (e.g, -20). Store the interval value as an index for the cell. Do an equal-to query to see if a cell is in an interval. So, for tmean =142 in the datastore, the 1-degree category index would be '14', meaning the actual value is in the range [14,15}. 

Do the same thing for as many different interval sizes make sense, each a power of 2 bigger than the last. So, continuing with the temperature example, create another index for a 2-degree intervals. The categories would be:

[-20, -18}, [-18,-16}, [-16,-14}...[38,40},[40,42}  (31 intervals)

For tmean = 142, the 2-degree category index would be '14', meaning the value is in the range [14,16}

The maximum number of interval categories needed to fully divide up the value space for a variable would be something like:

```
Formula for number of intervals given min/max values of a cell:
ceil(log2(ceil(max) - floor(min)))+1

Formula for index value:
floor(min_var_val + ((cell_val - (min_var_val)) / 2^n)))

Note: 2^n is the interval size
```

The same categories could be used for all temperatures, and another set for all precipitations, another for the altitudes, and another for the percentages.

Estimates for the maximum number of categories by variable type:
T [-90 to 60} 7 categories
P [0,14000} 14 categories
A [-424,8850} 14 categories
% [0,100} 6 categories

Several categories could easily be dropped from the precipitation and altitude categories, as it is unlikely people would want to know precipitation to within 1,2,4,8mm or altitude with within 1,2,4,8m.

Combinations of equality filters could allow refined ranges. For example, if you wanted tmean in the categories 13,14,and 15, you could issue a query:
4-degree interval = 12 and 1-degree interval != 12 and 1-degree interval != 16

13,14,15

This is because the 4-degree intervals for the tmean example above would be:

[-20, -16}, [-16,-12}, [-12,-8}...[36,40},[40,44}  (16 intervals)

And smallest interval size containing all of the sought values would be the 4-degree interval with index '12', meaning the value is in the range [12,16}.

**Comment by eightysteele** [comment]
Nice. It's going to take me a few to digest this though. In the meantime:

> The interval would be stored as its lower bound (e.g, -20). Store the interval value as an index for the cell. Do an equal-to query to see if a cell is in an interval. So, for tmean =142 in the datastore, the 1-degree category index would be '14', meaning the actual value is in the range [14,15}.

What would your `CellIndex` entity look like to support this? What would the query look like?

**Comment by tucotuco** [comment]
Example CellIndex for Cell having key=’2343-3434’ where tmean1=142, tmean2=173, alt=320 under the assumption that the range of tmean across all cells and time periods is [-19.4,40.6}, and the range of alt across all cells is [-454,8850}

```
CellIndex(model.Expando):
    name =model.StringProperty()

CellIndex(key_name=’2343-3434’, parent=Cell(key_name=’2343-3434’))
CellIndex.m1a = ‘14’ # [14,15}
CellIndex.m1b = ‘14’ # [14,16}
CellIndex.m1c = ‘12’ # [12,16}
CellIndex.m1d = ‘12’ # [12,20}
CellIndex.m1e = ‘12’ # [12,28}
CellIndex.m1f = ‘12’ # [12,44}

CellIndex.m2a = ‘17’  # [17,18}
CellIndex.m2b = ‘16’  # [16,18}
CellIndex.m2c = ‘16’  # [16,20}
CellIndex.m2d = ‘12’  # [12,20}
CellIndex.m2e = ‘12’  # [12,28}
CellIndex.m2f = ‘12’  # [12,44}

CellIndex.a1a = ‘320’ # [320,321} 2^0
CellIndex.a1b = ‘320’ # [320,322} 2^1
CellIndex.a1c = ‘318’ # [318,322} 2^2
CellIndex.a1d = ‘314’ # [314,322} 2^3
CellIndex.a1e = ‘314’ # [314,330} 2^4
CellIndex.a1f = ‘314’ # [314,346} 2^5
CellIndex.a1g = ‘314’  # [314,378} 2^6
CellIndex.a1h = ‘314’ # [314,442} 2^7
CellIndex.a1i = ‘314’ # [314,570} 2^8
CellIndex.a1j = ‘58’ # [58,570} 2^9
CellIndex.a1k = ‘-454’ # [-454,570} 2^10
CellIndex.a1l = ‘-454’ # [-454,1594} 2^11
CellIndex.a1m = ‘-454’ # [-454,3642} 2^12
CellIndex.a1n = ‘-454’ # [-454,7738} 2^13
```

‘m1’ signifies the variable m1 (alias for tmean1)
‘m2’ signifies the variable m2 (alias for tmean2)
‘a’ signifies the variable a (alias alt)

The trailing letter is the least-storage way of signifying the interval category:
a = 1-degree interval for temperatures
b = 2-degree interval for temp
c = 4-degree interval for temp
d = 8-degree interval for temp
e = 16-degree interval for temp
f = 32-degree interval for temp

Complete set of example intervals for t in [-20,40}
[-20, -19}, [-19,-18}, [-18,-17}...[39,40},[40,41} (61 1-degree intervals)
[-20, -18}, [-18,-16}, [-16,-14}...[38,40},[40,42} (31 2-degree intervals)
[-20, -16}, [-16,-12}, [-12,-8}...[36,40},[40,44} (16 4-degree intervals)
[-20, -12}, [-12,-4}, [-4,4}, [4,12}, [12,20}, [20,28}, [28,36}, [36,44} (8 8-degree intervals)
[-20, -4}, [-4,12}, [12,28}, [28,44} (4 16-degree intervals)
[-20, 12}, [12,44} (2 32-degree intervals)

Now, how would you use this? Well, to get all of the cells having tmean1 in the interval 14 to 15 degrees, simply query:

CellIndex.m1a==’14’  Boom

For all cells having tmean2 in the interval 10-20, query as

CellIndex.m2d==’12’ or CellIndex.m2b==’10’

where the first part returns all cells having values in the range [12,20} and the second, all cells in the range [10,12}

Using combinations in this way allows for querying on any range of values within the defined categories. Flexible. Storage less than for even a few constrained simple ranges.

There would be similar CellIndex values for every variable for which interval querying is desired. The worst case variables are altitude and precipitation, which have large integer ranges. These could be reduced by four categories by having intervals only for the multiples of ten in the native units.

So, for the altitude example above, 320m using only 10m resolution, the CellIndex values would look like this:

```
CellIndex.a1a = ‘32’ # [320,330}
CellIndex.a1b = ‘32’ # [320,340}
CellIndex.a1c = ‘30’ # [320,324}
CellIndex.a1d = ‘26’ # [260,340}
CellIndex.a1e = ‘18’ # [180,340}
CellIndex.a1f = ‘18’ # [180,500}
CellIndex.a1g = ‘18’ # [180,820}
CellIndex.a1h = ‘-46’ # [-460,820}
CellIndex.a1i = ‘-46’ # [-460,2210}
CellIndex.a1j = ‘-46’ # [-460,4660}
```

Any problems foreseen with this approach?

**Comment by tucotuco** [comment]
Intervals solution implemented and test with commit https://github.com/eightysteele/Spatial-Data-Library/commit/5ef033654480dabc5222234012f55ec77954c5a9. Needs to be folded into the search API.

**Comment by eightysteele** [comment]
Blocked on issue #32.
